### ISSUE TITLE
Any plan to implement datafusion go/c++ binding?

### ISSUE THREAD

**Issue by xiaofan-luan** [apache/arrow-datafusion#2485]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

Hi All, I'm from Milvus community(https://github.com/search?q=milvus) and we are a vector database written in Go/C++, I really love what datafusion is doing!
We store our scalar data in arrow, currently we copy arrow into vector or other format for executing predication.
Datafusion fits our need to filter out data from arrow, just curious if we have any plan to support datafusion in cpp~
Thanks!

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by houqp** [comment]
Hi @xiaofan-luan , I have been following milvus for a long time and would love to see a synergy between the two communities :D

It shouldn't be too hard to expose our high level interfaces through C ABI for C++ and Go. For example: http://bitboom.github.io/calling-rust-from-cpp. We might need to add a couple of small high level public APIs to help pass record batches between C++ and Rust through the Arrow C Data interface.

I don't know if anyone is actively working on the C binding for datafusion right now, but if there is enough interests, we could create a new repo at https://github.com/datafusion-contrib to facilitate the collaboration.

**Comment by Jimexist** [comment]
> I don't know if anyone is actively working on the C binding for datafusion right now, but if there is enough interests, we could create a new repo at https://github.com/datafusion-contrib to facilitate the collaboration.

indeed, that C API could also benefit Java binding as well, because it'll be much better than the current JNI approach (and also in future when Java panama project is landed)

**Comment by xiaofan-luan** [comment]
@houqp great! let me know if we decide to do so, we can help on evaluating the initial version and might help on contributing  it!

**Comment by houqp** [comment]
Now we just need to wait for someone from the community to lead this effort :D

**Comment by gedw99** [comment]
He will do it !!

**Comment by codeboyyong** [comment]
hello, any update on this topic? I am also interesting in how to call data fusion in go app.

**Comment by derekperkins** [comment]
Also interested in a Go integration

**Comment by adarshsrivastava11** [comment]
Since there is already https://github.com/datafusion-contrib/datafusion-c/tree/main, can we not directly use it in our go code? Do you have any tips or caveats to watch out for? @houqp

**Comment by erezrokah** [comment]
> Now we just need to wait for someone from the community to lead this effort :D

Let's say someone is interested in looking into implementing the Go bindings, what would be the best place to start?

**Comment by jayzhan211** [comment]
I guess the question is actually how to use rust code in golang, and I think there is no a de facto solution yet. I found a [post](https://blog.arcjet.com/calling-rust-ffi-libraries-from-go/#calling-the-rust-library-in-go) that maybe helpful.

**Comment by jakthom** [comment]
I've been playing, if others are interested - https://github.com/datafusion-contrib/datafusion-go
